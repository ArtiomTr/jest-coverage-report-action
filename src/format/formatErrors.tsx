import { toMarkdown } from 'mdast-util-to-markdown';

import { Errors } from '../components/Errors.js';

export const formatErrors = (errors: Array<Error>) => {
    return toMarkdown(
        <root>
            <Errors errors={errors} />
        </root>,
        {
            fences: true,
        }
    );
};
