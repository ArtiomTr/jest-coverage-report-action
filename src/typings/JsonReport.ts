export type JsonReport = {
    numFailedTestSuites: number;
    numFailedTests: number;
    numPassedTestSuites: number;
    numPassedTests: number;
    numPendingTestSuites: number;
    numPendingTests: number;
    numRuntimeErrorTestSuites: number;
    numTodoTests: number;
    numTotalTestSuites: number;
    numTotalTests: number;
    openHandles?: unknown[];
    snapshot: Snapshot;
    startTime: number;
    success: boolean;
    testResults?: TestResult[];
    wasInterrupted: boolean;
    coverageMap: CoverageMap;
};

export type Snapshot = {
    added: number;
    didUpdate: boolean;
    failure: boolean;
    filesAdded: number;
    filesRemoved: number;
    filesRemovedList?: unknown[];
    filesUnmatched: number;
    filesUpdated: number;
    matched: number;
    total: number;
    unchecked: number;
    uncheckedKeysByFile?: unknown[];
    unmatched: number;
    updated: number;
};

export type TestResult = {
    assertionResults?: AssertionResult[];
    endTime: number;
    message: string;
    name: string;
    startTime: number;
    status: string;
    summary: string;
};

export type AssertionResult = {
    ancestorTitles?: string[];
    failureMessages?: string[];
    fullName: string;
    location: Location;
    status: string;
    title: string;
};

export type Location = {
    column?: number;
    line: number;
};

export type Range = {
    start?: Location;
    end?: Location;
};

export type CoverageMap = Record<string, FileCoverage>;

export type FileCoverage = {
    path: string;
    statementMap: StatementMap;
    fnMap: FunctionMap;
    branchMap: BranchMap;
    s: HitMap;
    f: HitMap;
    b: ArrayHitMap;
};

export type StatementMap = Record<number, StatementCoverage>;

export type StatementCoverage = {
    start: Location;
    end: Location;
};

export type FunctionMap = Record<number, FunctionCoverage>;

export type FunctionCoverage = {
    name: string;
    decl: Range;
    loc: Range;
};

export type BranchMap = Record<number, BranchCoverage>;

export type BranchCoverage = {
    loc: Range;
    type: string;
    locations?: Range[];
};

export type HitMap = Record<number, number>;

export type ArrayHitMap = Record<number, number[]>;
