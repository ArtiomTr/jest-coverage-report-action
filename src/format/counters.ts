import { FileCoverage } from '../typings/JsonReport';

export const standardTotalCounter = (key: keyof FileCoverage) => (
    value: FileCoverage
) => Object.values(value[key]).length;

export const standardCoveredCounter = (key: keyof FileCoverage) => (
    value: FileCoverage
) => Object.values(value[key]).filter((hits) => hits > 0).length;

export const totalBranchesCounter = (value: FileCoverage) =>
    Object.values(value.b).reduce((acc, branch) => acc + branch.length, 0);

export const coveredBranchesCounter = (value: FileCoverage) =>
    Object.values(value.b).reduce(
        (acc, branch) => acc + branch.filter((hits) => hits > 0).length,
        0
    );

export const totalLinesCounter = (value: FileCoverage) => {
    const stats: Record<string, number> = getLineCoverage(value);
    return Object.keys(stats).length;
};

export const coveredLinesCounter = (value: FileCoverage) => {
    const stats: Record<string, number> = getLineCoverage(value);
    return Object.values(stats).filter((v) => !!v).length;
};

const getLineCoverage = (value: FileCoverage) => {
    const statementMap = value.statementMap;
    const statements = value.s;

    return Object.entries(statements).reduce((acc, [st, count]) => {
        const _st: number = parseInt(st);

        if (!statementMap[_st]) {
            return acc;
        }
        const { line } = statementMap[_st].start;
        const prevVal = acc[line];
        if (prevVal === undefined || prevVal < count) {
            acc[line] = count;
        }
        return acc;
    }, {} as Record<string, number>);
};
