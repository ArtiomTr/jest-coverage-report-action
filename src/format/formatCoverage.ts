import { parseDetails } from './details/parseDetails';
import { parseSummary } from './summary/parseSummary';
import { getFormattedCoverage } from './getFormattedCoverage';
import { JsonReport } from '../typings/JsonReport';

export const formatCoverage = (
    headReport: JsonReport | undefined,
    baseReport: JsonReport | undefined,
    threshold: number | undefined,
    hideDetails: boolean | undefined
): string => {
    if (headReport) {
        return getFormattedCoverage(
            parseSummary(headReport),
            baseReport ? parseSummary(baseReport) : undefined,
            parseDetails(headReport),
            baseReport ? parseDetails(baseReport) : undefined,
            threshold,
            hideDetails
        );
    }

    return '';
};
