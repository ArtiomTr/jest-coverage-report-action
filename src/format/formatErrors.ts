import { ActionError } from '../typings/ActionError';
import { getConsoleLink } from '../utils/getConsoleLink';
import { i18n } from '../utils/i18n';

const getNumberWidth = (index: number) => Math.floor(Math.log10(index));

export const formatErrors = (errors: Array<Error>) => {
    if (errors.length === 0) {
        return '';
    }

    if (errors.length === 1) {
        const error = errors[0];

        if (error instanceof ActionError) {
            return i18n(':x: {{ error }}', { error: error.toString() });
        }

        return i18n(':x: {{ unexpectedError }} \n```\n{{ error }}\n```', {
            error: error.toString(),
            unexpectedError: i18n('errors.unexpectedError', {
                consoleLink: getConsoleLink(),
            }),
        });
    }

    return (
        i18n('errors.multiple') +
        i18n('\n```\n{{ errors }}\n```\n', {
            errors: errors
                .map(
                    (error, index) =>
                        ` ${String(1 + index).padEnd(
                            1 + getNumberWidth(errors.length),
                            ' '
                        )} | ${error.toString()}`
                )
                .join('\n'),
        })
    );
};
