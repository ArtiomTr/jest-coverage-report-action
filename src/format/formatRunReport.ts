import { TestRunReport } from '../typings/Report';
import { createMarkdownSpoiler } from '../utils/createMarkdownSpoiler';

export const formatRunReport = (report: TestRunReport): string => {
    const parts = [`## ${report.title}`];
    if (report.failures) {
        parts.push(
            createMarkdownSpoiler({
                summary: report.summary,
                body: report.failures,
            })
        );
    } else {
        parts.push(report.summary);
    }
    return parts.join('\n');
};
