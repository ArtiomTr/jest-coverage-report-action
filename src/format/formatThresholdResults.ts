import { ThresholdResult } from '../typings/ThresholdResult';
import { i18n } from '../utils/i18n';

export const formatThresholdResults = (results: ThresholdResult[]): string => {
    return results
        .map(({ type, path, expected, received }) =>
            i18n(`thresholdFailures.${expected < 0 ? 'ones' : 'percents'}`, {
                path,
                type: i18n(type),
                ltype: i18n(type).toLowerCase(),
                expected: Math.abs(expected),
                coverage: received,
            })
        )
        .join('\n');
};
