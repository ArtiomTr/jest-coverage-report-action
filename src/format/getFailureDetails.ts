import stripAnsi from 'strip-ansi';

import { JsonReport } from '../typings/JsonReport';

export const getFailureDetails = ({ testResults }: JsonReport): string => {
    if (
        !testResults ||
        !testResults.some(
            ({ message, status }) => message.length > 0 && status !== 'passed'
        )
    ) {
        return '';
    }
    const wrapCode = (code: string) => '```\n' + code + '\n```';
    const codeBlocks = testResults
        .filter(({ status }) => status !== 'passed')
        .map(({ message }) => {
            const stripped = stripAnsi(message);
            if (stripped.trim().length === 0) {
                return '';
            }
            return wrapCode(stripped);
        })
        .filter(({ length }) => length > 0);
    return codeBlocks.join('\n---\n');
};
