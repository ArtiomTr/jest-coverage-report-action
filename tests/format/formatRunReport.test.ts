import { formatRunReport } from '../../src/format/formatRunReport';
import { getFailureDetails } from '../../src/format/getFailureDetails';
import { JsonReport } from '../../src/typings/JsonReport';

describe('formatRunReport', () => {
    it('should generate summary as markdown (successful tests)', () => {
        expect(
            formatRunReport({
                summary: '2 Tests Passed',
                title: 'Tests Passed',
            })
        ).toMatchSnapshot();
    });
    it('should generate summary as markdown (failure tests)', () => {
        expect(
            formatRunReport({
                summary: '2 Tests Passed, 1 Test Failed',
                title: 'Tests Failed',
                failures: getFailureDetails({
                    testResults: [
                        {
                            message: 'fail 1',
                        },
                        {
                            message: 'fail 2',
                        },
                    ],
                } as JsonReport),
            })
        ).toMatchSnapshot();
    });
});
