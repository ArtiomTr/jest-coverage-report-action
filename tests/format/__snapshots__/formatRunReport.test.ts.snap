// Vitest Snapshot v1, https://vitest.dev/guide/snapshot.html

exports[`formatRunReport > should generate summary as markdown (failure tests) 1`] = `
"## Tests Failed

<details><summary>2 Tests Passed, 1 Test Failed</summary>
<br/>

\`\`\`
fail 1
\`\`\`
---
\`\`\`
fail 2
\`\`\`

</details>
"
`;

exports[`formatRunReport > should generate summary as markdown (successful tests) 1`] = `
"## Tests Passed
2 Tests Passed"
`;
